File: astra-plugin-sdk-python/astra_plugin_sdk/testing/fixtures.py
"""Golden audio, config fuzz sets, and the pytest fixtures that hand them out.

This module is registered as a pytest plugin through the `pytest11` entry point
in `pyproject.toml`, so a plugin's tests get `astra_harness`, `astra_wire`,
`golden_pcm` and `fuzz_configs` with no `conftest.py`. Importing it directly
works too — nothing here needs pytest to exist.
"""

from __future__ import annotations

import math
import struct
from typing import Any, Callable, Iterator

from astra_plugin_sdk import limits
from astra_plugin_sdk.proto import plugin_pb2
from astra_plugin_sdk.types import AudioChunk

__all__ = [
    "golden_pcm_f32",
    "golden_pcm_i16",
    "wake_seed_burst",
    "backpressure_burst",
    "fuzz_configs",
    "FUZZ_CONFIGS",
]

#: Sample rate the daemon's voice pipeline runs at, everywhere.
SAMPLE_RATE = 16000

#: How long the wake-word seed buffer can be: the daemon replays the audio it
#: had already captured when the wake word fired, so the first thing a
#: recognizer sees is a burst of history, not live audio. `spec/limits.yaml`
#: sizes the channel for this exact case.
WAKE_SEED_SECONDS = 8.0

#: The daemon's batch size on the voice path.
BATCH_MS = 100


def golden_pcm_f32(
    seconds: float = 1.0,
    *,
    sample_rate: int = SAMPLE_RATE,
    frequency: float = 440.0,
    amplitude: float = 0.25,
) -> bytes:
    """Deterministic 32-bit float PCM — the format `SttProcess` carries.

    A sine, not noise or silence: silence is indistinguishable from a bug that
    sends nothing, and a random buffer makes a failure unreproducible. Same
    bytes on every machine and every run, so a test that asserts on a byte count
    or a hash keeps meaning the same thing.
    """
    frames = int(seconds * sample_rate)
    step = 2.0 * math.pi * frequency / sample_rate
    return struct.pack(f"<{frames}f", *(amplitude * math.sin(step * n) for n in range(frames)))


def golden_pcm_i16(
    seconds: float = 1.0,
    *,
    sample_rate: int = SAMPLE_RATE,
    frequency: float = 440.0,
    amplitude: float = 0.25,
) -> bytes:
    """The same signal as 16-bit signed PCM, for a plugin that wants that."""
    frames = int(seconds * sample_rate)
    step = 2.0 * math.pi * frequency / sample_rate
    peak = int(amplitude * 32767)
    return struct.pack(
        f"<{frames}h", *(int(peak * math.sin(step * n)) for n in range(frames))
    )


def _chunks(
    pcm: bytes, *, sample_rate: int, batch_ms: int, bytes_per_frame: int
) -> list[AudioChunk]:
    per_chunk = max(1, int(sample_rate * batch_ms / 1000) * bytes_per_frame)
    out = [
        AudioChunk(data=pcm[i : i + per_chunk], sample_rate=sample_rate)
        for i in range(0, len(pcm), per_chunk)
    ]
    if out:
        out[-1].is_last = True
    return out


def wake_seed_burst(
    *,
    seconds: float = WAKE_SEED_SECONDS,
    sample_rate: int = SAMPLE_RATE,
    batch_ms: int = BATCH_MS,
) -> list[AudioChunk]:
    """The wake-word seed dump, chunked the way the daemon sends it.

    ~8 seconds of already-captured audio, in 100 ms batches, arriving at once
    the instant the wake word fires. This is the burst `spec/limits.yaml` sizes
    the STT audio channel for, and it is what a recognizer that assumes audio
    arrives in real time gets wrong.
    """
    return _chunks(
        golden_pcm_f32(seconds, sample_rate=sample_rate),
        sample_rate=sample_rate,
        batch_ms=batch_ms,
        bytes_per_frame=4,
    )


def backpressure_burst(
    *, chunks: int = limits.STT_AUDIO_CHANNEL_CAPACITY + 100, frames: int = 160
) -> list[AudioChunk]:
    """More chunks than the SDK's audio queue can hold, each one identifiable.

    THE CONDITION THIS REPRODUCES. The SDK buffers inbound STT audio in a queue
    bounded by `limits.STT_AUDIO_CHANNEL_CAPACITY` (500). A smaller bound on
    either side of a two-channel bridge is the real capacity — which is how a
    32-slot SDK queue silently truncated every utterance past its first fraction
    against a 500-slot daemon channel, with no error anywhere. A correct SDK
    slows the sender; a broken one drops the tail.

    Each chunk's first four bytes are its index, so a test can assert not only
    that 600 chunks arrived but that they arrived *in order and complete*.
    """
    out = []
    for i in range(chunks):
        payload = struct.pack("<I", i) + bytes(frames * 4 - 4)
        out.append(AudioChunk(data=payload, sample_rate=SAMPLE_RATE))
    if out:
        out[-1].is_last = True
    return out


def chunk_index(chunk: AudioChunk | plugin_pb2.PluginAudioChunk) -> int:
    """Read back the index `backpressure_burst` stamped into a chunk."""
    data = chunk.data
    return struct.unpack("<I", data[:4])[0]


#: Config values a plugin will meet and usually has not thought about.
#:
#: Every entry has been a real bug somewhere: a user clearing a text field
#: leaves `""` and not the default; a number field arrives as a string from a
#: JSON form; a nested object arrives when the schema said scalar; and `None` is
#: what the daemon sends for a key the user has never touched. A plugin is
#: allowed to reject any of them — what it may not do is raise out of
#: `on_config_changed`, because that hook has nowhere to put a failure and the
#: plugin then runs on with the config it had.
FUZZ_CONFIGS: tuple[dict[str, Any], ...] = (
    {},
    {"unknown_key": "surprise"},
    {"api_key": ""},
    {"api_key": None},
    {"count": "12"},
    {"count": -1},
    {"count": 10**18},
    {"count": 1.5},
    {"enabled": "false"},
    {"nested": {"a": {"b": [1, 2, 3]}}},
    {"text": "ünïcödé — 🎧 \x00 \n\t"},
    {"text": "x" * 100_000},
    {"list": []},
    {"list": [None, {}, []]},
)


def fuzz_configs() -> tuple[dict[str, Any], ...]:
    """The fuzz set, as a tuple you can parametrize over."""
    return FUZZ_CONFIGS


# ── pytest fixtures (registered via the `pytest11` entry point) ──────────────

try:  # pragma: no cover — pytest is optional at runtime
    import pytest
except ImportError:  # pragma: no cover
    pytest = None  # type: ignore[assignment]


if pytest is not None:

    @pytest.fixture
    def astra_harness() -> Iterator[Callable[..., Any]]:
        """Build level-1 harnesses; every one is stopped when the test ends.

            def test_word_count(astra_harness):
                h = astra_harness(TextUtils(), config={"max_text_length": 20})
                assert h.call_tool("word_count", text="a b").json["words"] == 2

        A factory rather than a fixture returning one harness, because only the
        test knows which plugin, with which config.
        """
        from astra_plugin_sdk.testing.harness import Harness

        built: list[Harness] = []

        def make(plugin, *, config: dict | None = None, language: str = "en", **kwargs):
            harness = Harness(plugin, **kwargs)
            if config is not None:
                harness.with_config(config)
            harness.with_language(language)
            built.append(harness)
            return harness.start()

        try:
            yield make
        finally:
            for harness in built:
                harness.stop()

    @pytest.fixture
    def astra_wire() -> Iterator[Callable[..., Any]]:
        """Build level-2 wire harnesses; every one is stopped when the test ends."""
        from astra_plugin_sdk.testing.wire import WireHarness

        built: list[WireHarness] = []

        def make(plugin, **kwargs):
            harness = WireHarness(plugin, **kwargs)
            built.append(harness)
            return harness.start()

        try:
            yield make
        finally:
            for harness in built:
                harness.stop()

    @pytest.fixture
    def golden_pcm() -> bytes:
        """One second of deterministic 16 kHz f32 PCM."""
        return golden_pcm_f32(1.0)

    @pytest.fixture
    def wake_seed() -> list[AudioChunk]:
        """The 8-second wake-word seed dump, in 100 ms chunks."""
        return wake_seed_burst()

    @pytest.fixture(params=FUZZ_CONFIGS, ids=lambda c: str(sorted(c))[:40])
    def fuzz_config(request) -> dict[str, Any]:
        """One config from the fuzz set, parametrized over all of them."""
        return request.param

File: astra-plugin-sdk/src/testing/fixtures.rs
//! Golden inputs: audio the daemon really sends, chat events the firehose
//! really carries, and config payloads that really arrive.
//!
//! Everything here is **deterministic** — same bytes on every run, on every
//! machine, with no clock and no RNG. A fixture that varies is a test that
//! fails on someone else's laptop and nowhere else.

use crate::capability::SttOptions;
use crate::proto;

/// The sample rate the daemon's voice pipeline captures at, and therefore the
/// only rate an STT plugin is ever handed.
pub const SAMPLE_RATE: u32 = 16_000;

/// Samples in one captured batch: 100 ms at [`SAMPLE_RATE`].
///
/// The daemon batches the microphone at 1600 samples, which is why the
/// wake-word seed dump is ~80 chunks and not ~8.
pub const SAMPLES_PER_CHUNK: usize = 1_600;

/// Bytes in one chunk. The wire format is f32 little-endian PCM — 4 bytes a
/// sample — and a plugin that assumes i16 halves every duration it computes.
pub const BYTES_PER_CHUNK: usize = SAMPLES_PER_CHUNK * 4;

/// How long the wake-word seed buffer is, in captured chunks.
///
/// The daemon holds a rolling pre-roll so the utterance that *contains* the
/// wake word is transcribed rather than the tail after it, and dumps the whole
/// thing into the STT session the instant the word fires. ~8 s at 100 ms a
/// chunk. This is the burst that has to survive the trip.
pub const WAKE_SEED_CHUNKS: usize = 80;

/// One chunk of f32-LE PCM: a 440 Hz sine at `amplitude`, phase-continuous
/// from sample index `start`.
///
/// Deterministic and shaped like real audio — a plugin that computes a mean
/// amplitude (mock-stt does) gets a number that is stable and not zero.
pub fn pcm_chunk(start: usize, amplitude: f32) -> Vec<u8> {
    let mut out = Vec::with_capacity(BYTES_PER_CHUNK);
    for i in 0..SAMPLES_PER_CHUNK {
        let t = (start + i) as f32 / SAMPLE_RATE as f32;
        let v = amplitude * (std::f32::consts::TAU * 440.0 * t).sin();
        out.extend_from_slice(&v.to_le_bytes());
    }
    out
}

/// `chunks` chunks of continuous 440 Hz PCM.
pub fn pcm(chunks: usize) -> Vec<Vec<u8>> {
    (0..chunks)
        .map(|c| pcm_chunk(c * SAMPLES_PER_CHUNK, 0.25))
        .collect()
}

/// One second of speech-shaped audio: ten chunks. The ordinary case.
pub fn utterance() -> Vec<Vec<u8>> {
    pcm(10)
}

/// The wake-word seed dump: [`WAKE_SEED_CHUNKS`] chunks delivered back to back,
/// as fast as the transport will take them, before any live audio arrives.
///
/// This is the burst that exposed the 32-vs-500 channel mismatch: the daemon
/// pushed the whole pre-roll into a channel the SDK had sized at 32, and every
/// utterance reached the recognizer truncated to its first fraction.
pub fn wake_seed_burst() -> Vec<Vec<u8>> {
    pcm(WAKE_SEED_CHUNKS)
}

/// Exactly [`crate::limits::STT_AUDIO_CHANNEL_CAPACITY`] chunks — the largest
/// burst the contract says a plugin must absorb without the daemon's writer
/// blocking.
///
/// The count is read from the generated limit, never typed out: the number
/// lives in `spec/limits.yaml`, and a fixture with its own copy of it is
/// exactly the drift the spec exists to prevent.
pub fn channel_saturating_burst() -> Vec<Vec<u8>> {
    pcm(crate::limits::STT_AUDIO_CHANNEL_CAPACITY as usize)
}

/// The per-utterance decoding hints the daemon puts on the first chunk:
/// a language and the configured wake-word spellings as decoding bias.
pub fn stt_options() -> SttOptions {
    SttOptions {
        language: "en".into(),
        initial_prompt: "Astra, hey Astra".into(),
    }
}

// ── the chat firehose ────────────────────────────────────────────────────────

fn evt(seq: u64, event: proto::conversation_event_msg::Event) -> proto::ConversationEventMsg {
    proto::ConversationEventMsg {
        seq,
        timestamp: None,
        source_id: "test".into(),
        event: Some(event),
    }
}

/// One complete assistant turn as it arrives on `SubscribeChatEvents`: user
/// message, assistant start, two text deltas, a tool call and its result,
/// assistant complete.
///
/// Seqs are monotonic and start at 1, which is what a plugin that dedups by
/// seq needs in order to be testable at all.
pub fn firehose_events() -> Vec<proto::FirehoseEventMsg> {
    use proto::conversation_event_msg::Event as E;
    let conversation_id = "conv-1".to_string();
    let message_id = "msg-1".to_string();

    let events = vec![
        evt(
            1,
            E::UserMessage(proto::UserMessageEvt {
                message_id: "user-1".into(),
                content: "roll 2d20".into(),
                ..Default::default()
            }),
        ),
        evt(
            2,
            E::AssistantStart(proto::AssistantStartEvt {
                message_id: message_id.clone(),
            }),
        ),
        evt(
            3,
            E::AssistantTextDelta(proto::AssistantTextDeltaEvt {
                message_id: message_id.clone(),
                delta: "Rolling".into(),
            }),
        ),
        evt(
            4,
            E::ToolCallStart(proto::ToolCallStartEvt {
                message_id: message_id.clone(),
                call_id: "call-1".into(),
                name: "roll_dice".into(),
                arguments: r#"{"count":2,"sides":20}"#.into(),
            }),
        ),
        evt(
            5,
            E::ToolCallResult(proto::ToolCallResultEvt {
                call_id: "call-1".into(),
                output: "Rolled 2d20: [7, 20] = 27".into(),
                status: "completed".into(),
            }),
        ),
        evt(
            6,
            E::AssistantTextDelta(proto::AssistantTextDeltaEvt {
                message_id: message_id.clone(),
                delta: " — 27.".into(),
            }),
        ),
        evt(
            7,
            E::AssistantComplete(proto::AssistantCompleteEvt {
                message_id: message_id.clone(),
            }),
        ),
    ];

    events
        .into_iter()
        .map(|event| proto::FirehoseEventMsg {
            conversation_id: conversation_id.clone(),
            event: Some(event),
        })
        .collect()
}

/// A turn that ends in an error, which is the branch nobody writes a test for.
pub fn firehose_error_turn() -> Vec<proto::FirehoseEventMsg> {
    use proto::conversation_event_msg::Event as E;
    vec![
        proto::FirehoseEventMsg {
            conversation_id: "conv-2".into(),
            event: Some(evt(
                1,
                E::AssistantStart(proto::AssistantStartEvt {
                    message_id: "msg-2".into(),
                }),
            )),
        },
        proto::FirehoseEventMsg {
            conversation_id: "conv-2".into(),
            event: Some(evt(
                2,
                E::Error(proto::ErrorEvt {
                    message_id: "msg-2".into(),
                    content: "provider returned 429".into(),
                }),
            )),
        },
        // An event whose oneof is empty: a daemon newer than this SDK sending a
        // variant prost cannot name. A plugin must ignore it, not panic.
        proto::FirehoseEventMsg {
            conversation_id: "conv-2".into(),
            event: Some(proto::ConversationEventMsg {
                seq: 3,
                ..Default::default()
            }),
        },
    ]
}

// ── config ───────────────────────────────────────────────────────────────────

/// Config payloads that a plugin has to survive, each with the reason it is in
/// the list.
///
/// Not random: fuzzing a JSON parser tests serde. These are the payloads the
/// *daemon* actually produces, and every one of them has been a live plugin
/// bug — starting with `{}`, which is what a freshly installed plugin is sent
/// and which a config type without `#[serde(default)]` rejects.
pub fn config_fuzz() -> Vec<(&'static str, &'static str)> {
    vec![
        ("{}", "a fresh install: nothing has been configured yet"),
        ("", "the daemon has no stored config at all"),
        ("   ", "whitespace, from a hand-edited settings file"),
        ("null", "a settings row that was written as JSON null"),
        (
            r#"{"unknown_field":1}"#,
            "a setting from a newer version of this plugin",
        ),
        (
            r#"{"nested":{"a":[1,2,{"b":null}]}}"#,
            "arbitrary structure under an unknown key",
        ),
        (
            r#"{"default_sides":"twenty"}"#,
            "the right key with the wrong type — a hand-edited file",
        ),
        (
            r#"{"default_sides":-1}"#,
            "a negative where an unsigned is expected",
        ),
        (
            r#"{"default_sides":1e400}"#,
            "a number no f64 can hold",
        ),
        (
            r#"{"default_sides":6,"api_key":""}"#,
            "present but empty — the NotConfigured case, not a parse failure",
        ),
        (
            r#"{"text":"emoji 🎲 and \"quotes\" and \\ and  "}"#,
            "a string that has to survive JSON escaping",
        ),
        ("[]", "an array where an object belongs"),
        ("\"a string\"", "a scalar where an object belongs"),
        ("{", "truncated: a write that was interrupted"),
        (
            r#"{"a":1,}"#,
            "a trailing comma — legal in JSON5, not in JSON",
        ),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The fixture is bytes, not a description of bytes.
    #[test]
    fn pcm_is_f32_le_at_the_daemons_rate() {
        let chunk = &pcm(1)[0];
        assert_eq!(chunk.len(), BYTES_PER_CHUNK);
        let first = f32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]);
        assert_eq!(first, 0.0, "a sine starts at zero");
        // 440 Hz at 16 kHz: a quarter period is ~9.09 samples, so sample 9 is
        // near the peak.
        let peak = f32::from_le_bytes([chunk[36], chunk[37], chunk[38], chunk[39]]);
        assert!(peak > 0.24, "{peak}");
    }

    #[test]
    fn the_same_fixture_twice_is_the_same_bytes() {
        assert_eq!(pcm(3), pcm(3));
    }

    /// The burst is sized from the spec, so raising the limit in
    /// `spec/limits.yaml` widens the test instead of leaving it behind.
    #[test]
    fn the_saturating_burst_is_the_spec_number() {
        assert_eq!(
            channel_saturating_burst().len(),
            crate::limits::STT_AUDIO_CHANNEL_CAPACITY as usize
        );
        assert!(
            WAKE_SEED_CHUNKS < crate::limits::STT_AUDIO_CHANNEL_CAPACITY as usize,
            "the seed dump must fit in the channel with room for live audio — \
             that is what the limit is for"
        );
    }

    #[test]
    fn the_firehose_fixture_is_seq_monotonic() {
        let seqs: Vec<u64> = firehose_events()
            .iter()
            .map(|f| f.event.as_ref().unwrap().seq)
            .collect();
        assert_eq!(seqs, (1..=7).collect::<Vec<_>>());
    }
}
